diff --git a/novaclient/client.py b/novaclient/client.py
index b0d9aea1..f3bfd337 100644
--- a/novaclient/client.py
+++ b/novaclient/client.py
@@ -48,6 +48,9 @@ from novaclient import service_catalog
 from novaclient import utils
 
 
+REQ_ID_HEADER = 'X-OpenStack-Request-ID'
+
+
 class _ClientConnectionPool(object):
 
     def __init__(self):
@@ -184,12 +187,13 @@ class HTTPClient(object):
                  http_log_debug=False, auth_system='keystone',
                  auth_plugin=None, auth_token=None,
                  cacert=None, tenant_id=None, user_id=None,
-                 connection_pool=False):
+                 connection_pool=False, global_request_id=None):
         self.user = user
         self.user_id = user_id
         self.password = password
         self.projectid = projectid
         self.tenant_id = tenant_id
+        self.global_request_id = global_request_id
 
         self._connection_pool = (_ClientConnectionPool()
                                 if connection_pool else None)
@@ -389,6 +393,9 @@ class HTTPClient(object):
         kwargs.setdefault('headers', kwargs.get('headers', {}))
         kwargs['headers']['User-Agent'] = self.USER_AGENT
         kwargs['headers']['Accept'] = 'application/json'
+        if self.global_request_id:
+            kwargs['headers'].setdefault(REQ_ID_HEADER, self.global_request_id)
+
         if 'body' in kwargs:
             kwargs['headers']['Content-Type'] = 'application/json'
             kwargs['data'] = json.dumps(kwargs['body'])
@@ -740,7 +747,8 @@ def _construct_http_client(username=None, password=None, project_id=None,
                           os_cache=os_cache,
                           http_log_debug=http_log_debug,
                           cacert=cacert,
-                          connection_pool=connection_pool)
+                          connection_pool=connection_pool,
+                          global_request_id=kwargs.get('global_request_id'))
 
 
 def get_client_class(version):
