diff --git a/nova/context.py b/nova/context.py
index d93be7d8dd..0a30bf43a1 100644
--- a/nova/context.py
+++ b/nova/context.py
@@ -101,7 +101,8 @@ class RequestContext(context.RequestContext):
             show_deleted=kwargs.pop('show_deleted', False),
             request_id=request_id,
             resource_uuid=kwargs.pop('resource_uuid', None),
-            overwrite=overwrite)
+            overwrite=overwrite,
+            global_request_id=kwargs.pop('global_request_id', None))
         # oslo_context's RequestContext.to_dict() generates this field, we can
         # safely ignore this as we don't use it.
         kwargs.pop('user_identity', None)
