diff --git a/oslo_context/context.py b/oslo_context/context.py
index 44b2b36..d4fad22 100644
--- a/oslo_context/context.py
+++ b/oslo_context/context.py
@@ -45,7 +45,7 @@
     def __init__(self, auth_token=None, user=None, tenant=None, domain=None,
                  user_domain=None, project_domain=None, is_admin=False,
                  read_only=False, show_deleted=False, request_id=None,
-                 resource_uuid=None, overwrite=True):
+                 resource_uuid=None, overwrite=True, global_request_id=None):
         """Initialize the RequestContext
 
         :param overwrite: Set to False to ensure that the greenthread local
@@ -64,6 +64,7 @@
         if not request_id:
             request_id = generate_request_id()
         self.request_id = request_id
+        self.global_request_id = global_request_id
         if overwrite or not get_current():
             self.update_store()
 
@@ -88,8 +89,19 @@
                 'show_deleted': self.show_deleted,
                 'auth_token': self.auth_token,
                 'request_id': self.request_id,
+                'global_request_id': self.global_request_id,
                 'resource_uuid': self.resource_uuid,
                 'user_identity': user_idt}
+
+    @property
+    def global_id(self):
+        """Return a sensible value for global_id to pass on.
+
+        When we want to make a call with to another service, it's
+        important that we try to use global_request_id if available,
+        and fall back to the locally generated request_id if not.
+        """
+        return self.global_request_id or self.request_id
 
     @classmethod
     def from_dict(cls, ctx):
@@ -104,7 +116,8 @@
             read_only=ctx.get("read_only", False),
             show_deleted=ctx.get("show_deleted", False),
             request_id=ctx.get("request_id"),
-            resource_uuid=ctx.get("resource_uuid"))
+            resource_uuid=ctx.get("resource_uuid"),
+            global_request_id=ctx.get("global_request_id"))
 
 
 def get_admin_context(show_deleted=False):
